### PR TITLE
enh(api) New real time monitoring API for services and hosts

### PR DESCRIPTION
**Pull request by callapa**
## Description

This PR add a new version of the real time monitoring for services and hosts.
The technical stack use Symfony 4 and FOSRestBundle.

The new uri format is http[s]://_yourdomain.tld_/api/[**latest**|**v2**|**beta**]/monitoring/...

See API documentation for more information:

- in [HTML](https://github.com/centreon/centreon/blob/MON-4113/doc/API/centreon-api-v2.html)
- in [OpenAPI](https://github.com/centreon/centreon/blob/MON-4113/doc/API/centreon-api-v2.yaml)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [x] 19.10.x (master)

#### Community contributors & Centreon team

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [ ] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,2 @@
+APP_ENV=prod
+APP_SECRET=
```
Comment (kduret) [inline .env]: do we need to set a default secret ?

File: .gitignore [hunk 2]
```diff
@@ -20,6 +20,8 @@ www/include/core/menu/templates/*.tpl
 !www/img/media/.keep
 www/img/media/*
 www/template
+!filesGeneration/export/.keep
```
Comment (kduret) [inline .gitignore]: filesGeneration folder does not exist anymore

File: config/bootstrap.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+use Symfony\Component\Dotenv\Dotenv;
+
+require dirname(__DIR__).'/vendor/autoload.php';
```
Comment (kduret) [inline config/bootstrap.php]:
```suggestion
require dirname(__DIR__) . '/vendor/autoload.php';
```

File: config/bootstrap.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+// Load cached env vars if the .env.local.php file exists
+// Run "composer dump-env prod" to create it (requires symfony/flex >=1.2)
+if (file_exists(dirname(__DIR__).'/.env.local.php')
```
Comment (kduret) [inline config/bootstrap.php]:
```suggestion
if (file_exists(dirname(__DIR__) . '/.env.local.php')
```

File: config/bootstrap.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    && is_array($env = @include dirname(__DIR__).'/.env.local.php')
```
Comment (kduret) [inline config/bootstrap.php]:
```suggestion
    && is_array($env = @include dirname(__DIR__) . '/.env.local.php')
```

File: config/bootstrap.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+$_SERVER += $_ENV;
+$_SERVER['APP_ENV'] = $_ENV['APP_ENV'] = ($_SERVER['APP_ENV'] ?? $_ENV['APP_ENV'] ?? null) ?: 'dev';
+$_SERVER['APP_DEBUG'] = $_SERVER['APP_DEBUG'] ?? $_ENV['APP_DEBUG'] ?? 'prod' !== $_SERVER['APP_ENV'];
+$_SERVER['APP_DEBUG'] = $_ENV['APP_DEBUG'] =
+    (int) $_SERVER['APP_DEBUG']
+    || filter_var($_SERVER['APP_DEBUG'], FILTER_VALIDATE_BOOLEAN) ? '1' : '0';
+
+$conf_centreon = [];
+include_once dirname(__DIR__)."/config/centreon.config.php";
```
Comment (kduret) [inline config/bootstrap.php]:
```suggestion
include_once dirname(__DIR__) . "/config/centreon.config.php";
```

File: config/bootstrap.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+(new Dotenv())->populate($constants);
+
+include_once dirname(__DIR__)."/container.php";
```
Comment (kduret) [inline config/bootstrap.php]:
```suggestion
include_once dirname(__DIR__) . "/container.php";
```

File: config/bootstrap.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
+<?php
+
```
Comment (kduret) [inline config/bootstrap.php]: missing header

File: config/bundles.php [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<?php
+
```
Comment (kduret) [inline config/bundles.php]: missing header

File: config/packages/nelmio_cors.yaml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+nelmio_cors:
+    defaults:
+        allow_credentials: true
+        origin_regex: true
+        allow_origin: ['*']
+        allow_methods: ['GET', 'OPTIONS', 'POST', 'PUT', 'PATCH', 'DELETE']
+        allow_headers: ['Content-Type', 'Authorization', 'Origin','Accept','X-AUTH-TOKEN']
```
Comment (kduret) [inline config/packages/nelmio_cors.yaml]:
```suggestion
        allow_headers: ['Content-Type', 'Authorization', 'Origin', 'Accept', 'X-AUTH-TOKEN']
```

File: config/routes.yaml [hunk 11]
```diff
@@ -0,0 +1,3 @@
+#index:
```
Comment (kduret) [inline config/routes.yaml]: since we use annotation for routing, can we delete this file ?

File: container.php [hunk 12]
```diff
@@ -0,0 +1,79 @@
+<?php
+
```
Comment (kduret) [inline container.php]: missing header

File: doc/API/centreon-api-v2.yaml [hunk 13]
```diff
@@ -0,0 +1,958 @@
+openapi: 3.0.1
```
Comment (kduret) [inline doc/API/centreon-api-v2.yaml]: we need to split this file for better readability

File: public/index.php [hunk 14]
```diff
@@ -0,0 +1,27 @@
+<?php
+
```
Comment (kduret) [inline public/index.php]: missing header

File: src/Centreon/Domain/Entity/AuthenticationToken.php [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+    /**
+     * @return \DateTime
+     */
+    public function getGeneratedDate(): \DateTime
+    {
+        return $this->generatedDate;
+    }
+
+    public function isValid(): bool
```
Comment (kduret) [inline src/Centreon/Domain/Entity/AuthenticationToken.php]: missing comments

File: src/Centreon/Domain/Entity/EntityCreator.php [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+     * Create an entity and complete it according to the data array
+     *
+     * @param array $data Array that contains the data that will be used to complete entity
+     * @return mixed Return an instance of class according to the class name given into constructor
+     * @throws \Exception
+     */
+    public function createByArray(array $data)
+    {
+        if (!class_exists($this->className)) {
+            throw new \Exception('The class' . $this->className . ' does not exist');
```
Comment (kduret) [inline src/Centreon/Domain/Entity/EntityCreator.php]:
```suggestion
            throw new \Exception('The class ' . $this->className . ' does not exist');
```

File: src/Centreon/Domain/Pagination/RequestParameters.php [hunk 17]
```diff
@@ -0,0 +1,341 @@
+<?php
```
Comment (kduret) [inline src/Centreon/Domain/Pagination/RequestParameters.php]: Pagination directory should be renamed, cause it is not only pagination :)

File: src/Centreon/Domain/VersionHelper.php [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+     *
+     * @uses VersionHelper::regularizeDepthVersion() to harmonize the version numbers
+     * @uses version_compare after version numbers harmonization
+     *
+     * @param string $version1 First version to compare
+     * @param string $version2 Second version to compare
+     * @param string $operator Comparison operator (default: VersionHelper::EQUAL)
+     * @return bool Returns the comparison result
+     */
+    public static function compare(string $version1, string $version2, string $operator = self::EQUAL): bool
```
Comment (kduret) [inline src/Centreon/Domain/VersionHelper.php]: version_compare seems to already manage several depths
then, it is already possible to compare 4.1 with 4.1.2

File: src/Centreon/Infrastructure/DatabaseConnection.php [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+     * @param string $basename
+     * @param string $login
+     * @param string $password
+     * @param int $port
+     */
+    public function __construct(string $host, string $basename, string $login, string $password, int $port = 3306)
+    {
+        $dsn = "mysql:dbname={$basename};host={$host};port={$port}";
+        $options = array(
+            \PDO::MYSQL_ATTR_INIT_COMMAND => 'SET NAMES utf8',
```
Comment (kduret) [inline src/Centreon/Infrastructure/DatabaseConnection.php]: cause "use PDO;" at the top
```suggestion
            PDO::MYSQL_ATTR_INIT_COMMAND => 'SET NAMES utf8',
```

File: src/Centreon/Infrastructure/DatabaseConnection.php [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+            \PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION
```
Comment (kduret) [inline src/Centreon/Infrastructure/DatabaseConnection.php]: cause "use PDO;" at the top
```suggestion
            PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION
```

File: www/install/steps/templates/step7.tpl [hunk 21]
```diff
@@ -1,4 +1,4 @@
-{t}Currently installing database... please do not interrupt this process.{/t}<br/><br/>
+{t}Currently installing database and configuring cache... please do not interrupt this process.{/t}<br/><br/>
```
Comment (kduret) [inline www/install/steps/templates/step7.tpl]: IMO "generating cache" could be more appropriate

File: www/install/step_upgrade/step4.php [hunk 22]
```diff
@@ -48,7 +48,7 @@
 
 /*
 ** Get and check initial Centreon version.
-** Should be >= 2.8.0-beta1.
+** Should be >= 2.4.0.
```
Comment (kduret) [inline www/install/step_upgrade/step4.php]: 2.4.0 ?

File: src/Kernel.php [hunk 23]
```diff
@@ -0,0 +1,85 @@
+<?php
+
```
Comment (kduret) [inline src/Kernel.php]: missing header

File: src/EventSubscriber/CentreonEventSubscriber.php [hunk 24]
```diff
@@ -0,0 +1,259 @@
+<?php
+declare(strict_types=1);
```
Comment (kduret) [inline src/EventSubscriber/CentreonEventSubscriber.php]: missing header

File: src/EventSubscriber/CentreonEventSubscriber.php [hunk 25]
```diff
@@ -0,0 +1,259 @@
...
+                    continue;
+                }
+                foreach ($parameterValue as $subParameterName => $subParameterValues) {
+                    if (strpos($subParameterValues, '|') !== false) {
+                        $subParameterValues = explode('|', urldecode($subParameterValues));
+                        foreach ($subParameterValues as $value) {
+                            $search[RequestParameters::AGGREGATE_OPERATOR_OR][] = [$subParameterName => $value];
+                        }
+                    } else {
+                        $search[RequestParameters::AGGREGATE_OPERATOR_AND][$subParameterName] = $subParameterValues;
```
Comment (kduret) [inline src/EventSubscriber/CentreonEventSubscriber.php]: add urldecode

File: src/Centreon/Tests/Domain/Service/MonitoringServiceTest.php [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+use Centreon\Infrastructure\Repository\MonitoringRepositoryRDB;
+use PHPUnit\Framework\TestCase;
+
+class MonitoringServiceTest extends TestCase
+{
+
+
+    public function testFindOneHost()
+    {
+
```
Comment (kduret) [inline src/Centreon/Tests/Domain/Service/MonitoringServiceTest.php]: complete this unit test

File: src/Centreon/Tests/Domain/Service/MonitoringServiceTest.php [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    public function testFilterByContact()
+    {
```
Comment (kduret) [inline src/Centreon/Tests/Domain/Service/MonitoringServiceTest.php]: complete this unit test

File: src/Centreon/Infrastructure/Security/AuthenticationRepositoryRDB.php [hunk 28]
```diff
@@ -0,0 +1,280 @@
...
+     * @param array $details Data representing the authentication information
+     * @return AuthenticationToken Returns the new instance of the AuthenticationToken class
+     * @throws \Exception
+     */
+    private function createToken(array $details): AuthenticationToken
+    {
+        return new AuthenticationToken(
+            $details['token'],
+            (int) $details['contact_id'],
+            new \DateTime($details['generate_date'], new \DateTimeZone('Europe/Paris')),
```
Comment (kduret) [inline src/Centreon/Infrastructure/Security/AuthenticationRepositoryRDB.php]: we should use php timezone

File: src/Centreon/Infrastructure/Repository/SqlRequestParametersTranslator.php [hunk 29]
```diff
@@ -0,0 +1,314 @@
...
+    {
+        if ($aggregateOperator === RequestParameters::AGGREGATE_OPERATOR_AND) {
+            return 'AND';
+        } elseif ($aggregateOperator === RequestParameters::AGGREGATE_OPERATOR_OR) {
+            return 'OR';
+        }
+        throw new \Exception('Bad search operator');
+    }
+
+    private function translateSearchOperator(string $operator): string
```
Comment (kduret) [inline src/Centreon/Infrastructure/Repository/SqlRequestParametersTranslator.php]: add comments

File: src/Centreon/Infrastructure/Repository/SqlRequestParametersTranslator.php [hunk 30]
```diff
@@ -0,0 +1,314 @@
...
+            ? '(' . $databaseQuery . ')'
+            : $databaseQuery;
+    }
+
+    public function getRequestParameters(): RequestParametersInterface
+    {
+        return $this->requestParameters;
+    }
+
+    public function translatePaginationToSql(): ?string
```
Comment (kduret) [inline src/Centreon/Infrastructure/Repository/SqlRequestParametersTranslator.php]: add comments

File: src/Centreon/Infrastructure/Repository/SqlRequestParametersTranslator.php [hunk 31]
```diff
@@ -0,0 +1,314 @@
...
+            } else {
+                $databaseQuery .= $databaseSubQuery;
+            }
+        }
+        return count($search) > 1
+            ? '(' . $databaseQuery . ')'
+            : $databaseQuery;
+    }
+
+    public function getRequestParameters(): RequestParametersInterface
```
Comment (kduret) [inline src/Centreon/Infrastructure/Repository/SqlRequestParametersTranslator.php]: add comments

File: src/Centreon/Domain/VersionHelper.php (hunk 18 above)
Comment (callapa) [reply src/Centreon/Domain/VersionHelper.php]: Because there is a regularization in the version depth.
ex: 4 == 4.0

File: src/Kernel.php (hunk 23 above)
Comment (callapa) [reply src/Kernel.php]: File from Symfony, no Centreon licence.

File: src/Centreon/Infrastructure/Repository/SqlRequestParametersTranslator.php [hunk 32]
```diff
@@ -0,0 +1,329 @@
...
+    {
+        $whereQuery = '';
+        $search = $this->requestParameters->getSearch();
+        if (!empty($search) && is_array($search)) {
+            $this->searchValues = [];
+            $whereQuery .= $this->createDatabaseQuery($search);
+        }
+        return !empty($whereQuery) ? ' WHERE ' . $whereQuery : null;
+    }
+
```
Comment (loiclau) [inline src/Centreon/Infrastructure/Repository/SqlRequestParametersTranslator.php]: comment

File: src/EventSubscriber/CentreonEventSubscriber.php [hunk 33]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+    /**
+     * Use to update the api version into all responses
+     *
+     * @param FilterResponseEvent $event
+     */
+    public function addApiVersion(FilterResponseEvent $event)
+    {
+        $apiVersion = '1.0';
```
Comment (loiclau) [inline src/EventSubscriber/CentreonEventSubscriber.php]: 1 or 2 ?

File: tests/php/Centreon/Domain/Monitoring/MonitoringServiceTest.php [hunk 34]
```diff
@@ -0,0 +1,482 @@
+<?php
+
```
Comment (loiclau) [inline tests/php/Centreon/Domain/Monitoring/MonitoringServiceTest.php]: licence

File: tests/php/Centreon/Domain/VersionHelperTest.php [hunk 35]
```diff
@@ -0,0 +1,32 @@
+<?php
+
```
Comment (loiclau) [inline tests/php/Centreon/Domain/VersionHelperTest.php]: licence

File: www/install/steps/process/generationCache.php [hunk 36]
```diff
@@ -0,0 +1,62 @@
+<?php
+
```
Comment (loiclau) [inline www/install/steps/process/generationCache.php]: licence

File: src/Security/TokenAPIAuthenticator.php [hunk 37]
```diff
@@ -0,0 +1,276 @@
...
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\AuthenticationExpiredException;
+use Symfony\Component\Security\Core\Exception\CredentialsExpiredException;
+use Symfony\Component\Security\Core\Exception\TokenNotFoundException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Guard\AbstractGuardAuthenticator;
+
```
Comment (loiclau) [inline src/Security/TokenAPIAuthenticator.php]: comment

File: src/Security/SessionAPIAuthenticator.php [hunk 38]
```diff
@@ -0,0 +1,281 @@
...
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\SessionUnavailableException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Guard\AbstractGuardAuthenticator;
+
```
Comment (loiclau) [inline src/Security/SessionAPIAuthenticator.php]: comment

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 39]
```diff
@@ -2847,12 +2847,9 @@ msgid "Installation"
 msgstr "Installation"
 
 #: centreon-web/www/install/steps/step7.php:45
-msgid ""
-"Currently installing database... please do not interrupt this process.<br/"
-"><br/>"
-msgstr ""
-"Installation des bases de données en cours... ne pas interrompre le "
-"processus.<br/><br/>"
+msgid "Currently installing database and configuring cache... please do not interrupt this process."
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po]: generating instead of configuring

File: lang/pt_BR.UTF-8/LC_MESSAGES/messages.po [hunk 40]
```diff
@@ -13053,8 +13053,8 @@ msgstr "Sobrenome"
 msgid "Congratulations, you have successfully installed Centreon!"
 msgstr "Parabéns, o Centreon foi instalado com sucesso!"
 
-msgid "Currently installing database... please do not interrupt this process.<br/><br/>"
-msgstr "Efetuando a instalação do banco de dados ... por favor não interrompa este processo.<br/><br/>"
+msgid "Currently installing database and configuring cache... please do not interrupt this process.<br/><br/>"
```
Comment (kduret) [inline lang/pt_BR.UTF-8/LC_MESSAGES/messages.po]: generating instead of configuring

File: lang/pt_PT.UTF-8/LC_MESSAGES/messages.po [hunk 41]
```diff
@@ -13054,8 +13054,8 @@ msgstr "Apelido"
 msgid "Congratulations, you have successfully installed Centreon!"
 msgstr "Parabéns, o Centreon foi instalado com sucesso!"
 
-msgid "Currently installing database... please do not interrupt this process.<br/><br/>"
-msgstr "Efetuando a instalação do banco de dados ... por favor não interrompa este processo.<br/><br/>"
+msgid "Currently installing database and configuring cache... please do not interrupt this process.<br/><br/>"
```
Comment (kduret) [inline lang/pt_PT.UTF-8/LC_MESSAGES/messages.po]: generating instead of configuring

**Pull request merged**